### ISSUE TITLE
YAML extra-vars on the command line are not being parsed as YAML

### ISSUE THREAD

**Issue by russellcardullo** [ansible/ansible#38415]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
extra-vars

##### ANSIBLE VERSION
```
ansible 2.5.0
```

##### OS / ENVIRONMENT

Mac OS 10.12.6

##### SUMMARY
Passing in extra-vars as YAML on the command line is not interpreted as YAML.

##### STEPS TO REPRODUCE

Using the playbook `arcade.yml`:
```yaml
- hosts: all
  gather_facts: no
  tasks:
    - debug: var=pacman
```
Run the command passing in extra-vars on the command line as per the [documentation](http://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#passing-variables-on-the-command-line)
```
ansible-playbook arcade.yml --extra-vars '
pacman: mrs
ghosts:
- inky
- pinky
- clyde
- sue'
```

##### EXPECTED RESULTS
I expected that the variables passed as YAML would be read and output something like this:

```
ok: [localhost] => {
    "pacman": "mrs"
}
```

##### ACTUAL RESULTS
```
PLAY [all] *********************************************************************************************************************************************************************************************************************************************************************************

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "pacman": "VARIABLE IS NOT DEFINED!"
}

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0
```

I found that the YAML parsing is only triggered if the first character is in the set `[{` from the code [here](https://github.com/ansible/ansible/blob/b794fa4fe74c9063346dfed32727298aeecb2e8d/lib/ansible/utils/vars.py#L131-L133).  I was able to get this working by surrounding my YAML with braces like so:

```
ansible-playbook -c local -i localhost, arcade.yml --extra-vars '{
pacman: mrs,
ghosts:
inky,
pinky,
clyde,
sue}'
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ghost** [comment]
Since this case is falling in a condition where "{" and "[" chars are valid for JSON same as for YAML, I wonder if using "---" would make sense as an indicator of YAML syntax; then a case like:

```python          
            elif (
                extra_vars_opt and
                any((extra_vars_opt[0] in u'[{', extra_vars_opt.replace('\n', '')[:3] == u'---')) 
                #optional remove '\n' for the '---' evaluation only
            ):
```

This of course would need an update both in the code and documentation.
A test of this idea:

```
(venv) [monchitos82@ns1 vars]$ ansible-playbook arcade.yml --extra-vars '
---
pacman: mrs
ghosts:
- inky
- pinky
- clyde
- sue' -c local -vvvv
ansible-playbook 2.5.0
  config file = None
  configured module search path = ['/home/monchitos82/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/monchitos82/ansible_tests/venv/lib/python3.6/site-packages/ansible
  executable location = /home/monchitos82/ansible_tests/venv/bin/ansible-playbook
  python version = 3.6.4 (default, Dec 19 2017, 14:48:12) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
No config file found; using defaults
setting up inventory plugins
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

Loading callback plugin default of type stdout, v2.0 from /home/monchitos82/ansible_tests/venv/lib/python3.6/site-packages/ansible/plugins/callback/default.py

PLAYBOOK: arcade.yml *******************************************************************************************************************
1 plays in arcade.yml

PLAY [localhost] ***********************************************************************************************************************
META: ran handlers

TASK [debug] ***************************************************************************************************************************
task path: /home/monchitos82/ansible_tests/vars/arcade.yml:4
ok: [localhost] => {
    "pacman": "mrs"
}
META: ran handlers
META: ran handlers

PLAY RECAP *****************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0 
```

**Comment by wknapik** [comment]
Just ran into this myself ;/ A smaller example to reproduce the problem:
```
ansible localhost -m debug -a var=foo -e 'foo: 1'
```

**Comment by ghost** [comment]
Yep, the proposal seems to fix this too, not sure if the syntax is something we would like to see, I need some feedback here, I can wait on my PR since I identified 3 possibilities here (the old JSON is YAML but YAML is not JSON):

1. You tell Ansible you are using YAML (my proposal)
2. You tell Ansible you are **not** using YAML (this is good too!)
3. You let Ansible parse the whole statement and if it fails opt for something else (nay)

Without "proposed fix"
```bash
$ ansible localhost -m debug -a var=foo -e 'foo: 1'
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: Could not match supplied host pattern, ignoring: all

 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS => {
    "changed": false,
    "foo": "VARIABLE IS NOT DEFINED!"
}
```

Using "proposed fix"
```bash
$ ansible localhost -m debug -a var=foo -e '--- foo: 1' 
[WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

localhost | SUCCESS => {
    "foo": 1
}
```

**Comment by wknapik** [comment]
I don't think that's valid yaml:
```
--- foo: 1
```

But you can always call:
```
ansible localhost -m debug -a var=foo -e $'---\nfoo: 1'
```

**Comment by ghost** [comment]
Right, the condition pretty much will look for an indicator, not for parsing your whole statement. Adding the `\n` is not something I'd love to have (more slicing, more splitting) but this one works:
```
$ ansible localhost -m debug -a var=foo -e '---
foo: 1'
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

localhost | SUCCESS => {
    "foo": 1
}
```
Anyway I think the core maintainers opinion matters here, so I'll hold for them, unless you want to take the proposal and make the PR yourself (and see what happens? ¯\\_(ツ)_/¯)

**Comment by ghost** [comment]
Changed my mind about two things: (1) only a PR will draw the needed attention and (2) a YAML parser is the safest path here.

**Comment by sivel** [comment]
The source of this problem is invalid documentation that was added in https://github.com/ansible/ansible/pull/31433 indicating that YAML syntax on the CLI was acceptable.

We have decided against accepting multiline (non-JSONified YAML) on the command line.

As such, the solution to this issue is to pass JSON or JSONified YAML, or pass a file to extra vars.

I will work to remove invalid examples from the documentation.
